Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Analytics.hs

```diff
@@ -0,0 +1,593 @@
+  enabledCount <- CDI.getEnabledDriverCountByDriverIds driverIds (UTCTime fromDay 0) (UTCTime toDay 86399)
+  gt1 <- CDaily.countDriversWithNumRidesGreaterThan1Between driverIds fromDay toDay
+  gt10 <- CDaily.countDriversWithNumRidesGreaterThan10Between driverIds fromDay toDay
+  gt50 <- CDaily.countDriversWithNumRidesGreaterThan50Between driverIds fromDay toDay
+  pure (appCount, enabledCount, gt1, gt10, gt50)
+
+-- | Fallback to ClickHouse for period analytics and update Redis keys for the period
+fallbackToClickHouseAndUpdateRedisForPeriod ::
+  ( MonadFlow m,
+    EsqDBFlow m r,
+    CacheFlow m r,
+    Redis.HedisFlow m r,
+    HasField "serviceClickhouseCfg" r CH.ClickhouseCfg,
+    HasField "serviceClickhouseEnv" r CH.ClickhouseEnv
+  ) =>
+  Text ->
+  Text ->
+  Text ->
+  Text ->
+  Text ->
+  Text ->
+  Period ->
+  Day ->
+  Day ->
+  m (Int, Int, Int, Int, Int)
+fallbackToClickHouseAndUpdateRedisForPeriod operatorId kTotalApplicationCount kDriverEnabled kGtOne kGtTen kGtFifty period fromDay toDay = do
+  logTagInfo "Fallback to ClickHouse (period)" "true"
+  (appCount, enabledCount, gt1, gt10, gt50) <- fallbackComputePeriodOperatorAnalytics operatorId fromDay toDay
+  now <- getCurrentTime
+  -- Determine expiration time based on period
+  let expireTime = case period of
+        Today -> calculateSecondsUntilEndOfDay now
+        Yesterday -> calculateSecondsUntilEndOfDay now
+        ThisWeek -> calculateSecondsUntilEndOfWeek now
+        LastWeek -> calculateSecondsUntilEndOfWeek now
+        ThisMonth -> calculateSecondsUntilEndOfMonth now
+        LastMonth -> calculateSecondsUntilEndOfMonth now
+
+  -- write back to Redis using setNx with expiration
+  void $ Redis.setExp kTotalApplicationCount appCount expireTime
+  void $ Redis.setExp kDriverEnabled enabledCount expireTime
+  void $ Redis.setExp kGtOne gt1 expireTime
+  void $ Redis.setExp kGtTen gt10 expireTime
+  void $ Redis.setExp kGtFifty gt50 expireTime
+  pure (appCount, enabledCount, gt1, gt10, gt50)
+
+inferPeriod :: UTCTime -> UTCTime -> UTCTime -> Maybe Period
+inferPeriod now from to =
+  let (y, m, d) = toGregorian (utctDay now)
+      todayStart = UTCTime (fromGregorian y m d) 0
+      -- week starts on Monday
+      dow = calculateDayNumberOfWeek now
+      thisWeekStart = addUTCTime (negate (fromIntegral dow * 86400)) todayStart
+      firstOfMonth = UTCTime (fromGregorian y m 1) 0
+      yesterdayStart = addUTCTime (-86400) todayStart
+      lastWeekStart = addUTCTime (-7 * 86400) thisWeekStart
+      lastMonthStart = UTCTime (addGregorianMonthsClip (-1) (fromGregorian y m 1)) 0
+   in if from == todayStart && to <= now && to >= todayStart
+        then Just Today
+        else
+          if from == thisWeekStart && to <= now && to >= todayStart
+            then Just ThisWeek
+            else
+              if from == firstOfMonth && to <= now && to >= todayStart
```

Comment:
I guess its better to compare `Day` not `UTCTime`, then day time will not affect